### PR TITLE
Fused optimizer and gradient clipper using apex

### PR DESCRIPTION
**Pull request by slayton58**
Significant (40ms / iter for XLNet squad finetuning) performance increase.

Also adds fused grad clipping, gives further ~30ms / iter saving in the same XLNet squad case. 

Redefines the `AdamW` implementation such that the existing code will be used if apex's multi_tensor_apply code isn't available, and should drop-in speedup all existing scripts using `AdamW`. Also abstracts the gradient clipping (in order to keep run scripts concise and move apex-specific logic into `optimizations.py`)

### CODE REVIEW

**Comment by BramVanroy** [comment]
Does this apply that apex's FusedAdam is a fused implementation of AdamW rather than Adam (vanilla)?

It might be nice to first try to import [torch's native AdamW](https://pytorch.org/docs/stable/_modules/torch/optim/adamw.html#AdamW) (from 1.2), and if not available fallback to the transformers implementation.

**Comment by slayton58** [comment]
https://github.com/huggingface/transformers/pull/1516/files#diff-59de7b854fbd60c6ba87f68027a2db36R208 enables AdamW support in the `FusedAdam` optimizer. 

I see you've done the work to check for native PyT implementation & defining if necessary, it should be an easy rebase & resolve if/when these individual PRs get merged.

**Comment by BramVanroy** [comment]
Oh, my bad. I hadn't noticed this `adam_w_mode=True` in Apex's Adam before. Good to know!

**Comment by thomwolf** [comment]
Updated to clip gradient outside of gradient accumulation inner loop as we are trying to do now.

We should update the other training scripts as well (`run_glue` for instance). I think it may be soon time to refactor and gather the common portions of the examples (which are numerous) so we spend less time synchronizing them.

What do you think @LysandreJik?

**Comment by LysandreJik** [comment]
I agree that some scripts definitely need some refactoring and that having shared pieces of code like that gradient clipping seems like the way to go.
